src/Common/CoTask.h

```diff
@@ -0,0 +1,621 @@
+    details::SyncWaitTask::run(*this, result_holder, exception, event);
+    event.wait();
+    if (exception)
+        std::rethrow_exception(exception);
+    if constexpr (!std::is_void_v<ResultType>)
+        return std::move(*result_holder.getResultPtr());
+}
+
+template <typename ResultType>
+ResultType Task<ResultType>::syncRun(const Scheduler & scheduler) &&
+{
+    if constexpr (std::is_void_v<ResultType>)
+        std::move(*this).run(scheduler).syncWait();
+    else
+        return std::move(*this).run(scheduler).syncWait();
+}
+
+template <typename ResultType>
+class Task<ResultType>::PromiseBase
+{
+public:
+    void return_value(ResultType result_) { await_control->setResult(std::move(result_)); }
+protected:
+    std::shared_ptr<typename Task<ResultType>::AwaitControl> await_control;
+};
+
+template <>
+class Task<void>::PromiseBase
+{
+public:
+    void return_void() {}
+protected:
+    std::shared_ptr<typename Task<>::AwaitControl> await_control;
+};
+
+/// Represents the `promise_type` for coroutine tasks.
+template <typename ResultType>
+class Task<ResultType>::Promise : public PromiseBase
+{
+public:
+    Promise() { this->await_control = std::make_shared<AwaitControl>(std::coroutine_handle<Promise>::from_promise(*this)); }
+
+    Task get_return_object() { return Task(getAwaitControl()); }
+
+    Awaiter run(const Scheduler & scheduler_) { return getAwaitControl()->run(scheduler_); }
+
+    auto initial_suspend() { return std::suspend_always{}; }
+
+    class FinalSuspend;
+    auto final_suspend() noexcept { return FinalSuspend{getAwaitControl()}; }
+
+    void unhandled_exception() { getAwaitControl()->setException(std::current_exception()); }
+
+    auto await_transform(TaskRunParams)
+    {
+        struct Getter : std::suspend_never
+        {
+            TaskRunParams params;
+            TaskRunParams await_resume() const noexcept { return params; }
+        };
+        return Getter{{}, getAwaitControl()->getRunParams()};
+    }
+
+    auto await_transform(IsCancelled)
+    {
+        struct Getter : std::suspend_never
+        {
+            bool is_cancelled;
+            IsCancelled await_resume() const noexcept { return {is_cancelled}; }
+        };
+        return Getter{{}, static_cast<bool>(getAwaitControl()->getCancelStatus()->isCancelled())};
+    }
+
+    auto await_transform(StopIfCancelled)
+    {
+        struct Getter : std::suspend_never
+        {
+            std::exception_ptr cancel_exception;
+            void await_resume() const
+            {
+                if (cancel_exception)
+                    std::rethrow_exception(cancel_exception);
+            }
+        };
+        return Getter{{}, getAwaitControl()->getCancelStatus()->getCancelException()};
+    }
+
+    template <typename AwaiterType>
+    auto await_transform(AwaiterType && awaiter, typename std::enable_if_t<details::is_awaiter_type_v<std::remove_cvref_t<AwaiterType>>> * = nullptr)
```

Comment:
Concepts would be simpler and not dependent on boost
https://godbolt.org/z/6b5x3Wo5o
```suggestion
    requires requires(F f) { {f.await_ready()} -> std::same_as<bool>; }
    auto await_transform(AwaiterType && awaiter)
```